plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs

```diff
@@ -84,9 +92,7 @@ type family ToCostingType n where
 data BuiltinRuntime val =
     forall n. BuiltinRuntime
         (RuntimeScheme n)
-        ~(ToRuntimeDenotationType val n)  -- Must be lazy, because we don't want to compute the
-                                          -- denotation when it's fully saturated before figuring
-                                          -- out what it's going to cost.
+        (ToRuntimeDenotationType val n)
```

Comment:
Do you have an example of this value being unnecessarily forced before? I didn't spot any.

Reply:
It would be very wrong to let people run what they don't have a budget for, so we have tests for that. If you remove that tilda on `master` and run `cabal test plutus-core-test`, you'll get

```
     ExpensiveSucc:                                          FAIL
        untyped-plutus-core/test/Evaluation/Builtins/Definition.hs:200:
        expected: Right (Right EvaluationFailure)
         but got: Right (Left BuiltinErrorCall)
        Use -p '/ExpensiveSucc/' to rerun this test only.
      ExpensivePlus:                                          FAIL
        untyped-plutus-core/test/Evaluation/Builtins/Definition.hs:230:
        expected: Right (Right EvaluationFailure)
         but got: Right (Left BuiltinErrorCall)
        Use -p '/ExpensivePlus/' to rerun this test only.
```

Reply:
Yeah but I was wondering why `Lazy` is helpful. The reason you added `Lazy` is because this value was unnecessarily forced to WHNF somewhere - right?

Reply:
Ah, yeah, see the [competing](https://github.com/input-output-hk/plutus/pull/4584) PR that removes the unnecessary forcing. Somehow it was worse than the `Lazy` approach.

Reply:
Well maybe I need some more coffee :joy: but the competing PR simply avoids constructing and destructing `BuiltinRuntime`. But since the `~(ToRuntimeDenotationType val n)` field is lazy, constructing and destructing `BuiltinRuntime` doesn't force it, does it?

Reply:
> the competing PR simply avoids constructing and destructing BuiltinRuntime. But since the `~(ToRuntimeDenotationType val n)` field is lazy

Constructing and deconstructing is what forces us to have that tilda there. The competing PR avoids the constructing and deconstructing, which allows it to remove the tilda (the first file in the diff).

Reply:
Does that mean the saving in the competing PR didn't come from avoiding unnecessary forcing (since there wasn't unnecessary forcing before either), it came from avoiding the construction and destruction? Then why would this PR help since it doesn't avoid construction/destruction, but just replace the tilda with `Lazy`?

Reply:
> it came from avoiding the construction and destruction?

Avoiding the construction and deconstruction is why the competing PR should've been faster than the `Lazy` approach. However it's not the reason why the two PR exist. The point is that while we want to avoid premature forcing of the result, we don't want to avoid forcing of every intermediate partial application as keeping those thunks likely costs us some. I.e. we want to be lazy only for a fully saturated builtin and not for any partial application.

Reply:
Ah Ok, thanks, now I get it :joy: